nixos/modules/hardware/fwenv.nix

```diff
@@ -0,0 +1,65 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.hardware.fwenv;
+in
+
+{
+  options.hardware.fwenv = {
+    enable = lib.mkEnableOption (lib.mdDoc ''
+      U-Boot environment access configuration.
+
+      Use this to setup access to the U-Boot's environment variables with
+      generic ubootEnvTools package.
+    '');
+    package = lib.mkOption {
+      type = lib.types.package;
+      default = pkgs.ubootEnvTools;
+      defaultText = lib.literalExpression "pkgs.ubootEnvTools";
+      description = lib.mdDoc ''
+        Package providing U-Boot's environment tools (`fw_printenv` and
+        `fw_setenv`) that is added to `environment.systemPackages`.
+      '';
+    };
+    device = lib.mkOption {
+      type = lib.types.str;
+      description = lib.mdDoc "Device with U-Boot environment";
+    };
+    offset = lib.mkOption {
+      type = lib.types.ints.unsigned;
+      default = 0;
+      description = lib.mdDoc "Offset in the device to the environment";
+    };
+    size = lib.mkOption {
+      type = lib.types.ints.positive;
+      description = lib.mdDoc "Environment size";
+    };
+    secsize = lib.mkOption {
+      type = with lib.types; nullOr ints.positive;
+      default = null;
+      description = lib.mdDoc "Flash sector size.";
+    };
+    numsec = lib.mkOption {
+      type = with lib.types; nullOr ints.positive;
+      default = null;
+      description = lib.mdDoc "Number of sectors";
+    };
```

Comment:
We probably want to use a more nuanced type than just int.

People might want to use e.g. `"0x4000"` rather than `16384` in their config. There is no other literal forms than base 10 for numbers in Nix.

The type could be an either `int.positive` or a regex for matching strings like `0x[a-fA-F0-9]+`.

Reply:
Nix have float support, but not hexadecimal? What??

Reply:
```
~ $  nix repl
Welcome to Nix 2.15.1. Type :? for help.

nix-repl> 0x1234
error: undefined variable 'x1234'
```

Yeah :/.

Reply:
Let's use `strMatching` here. It would be better to have it as a hexadecimal integer, but since that is not supported in Nix the string is really our next best choice.